src/oidc_child/oidc_child_curl.c

```diff
@@ -0,0 +1,462 @@
+
+done:
+    talloc_free(post_data);
+    talloc_free(error_description);
+    curl_slist_free_all(headers);
+    curl_easy_cleanup(curl_ctx);
+    return ret;
+}
+
+errno_t get_openid_configuration(struct devicecode_ctx *dc_ctx,
+                                        const char *issuer_url)
+{
+    int ret;
+    char *uri = NULL;
+    bool has_slash = false;
+
+    if (issuer_url[strlen(issuer_url) - 1] == '/') {
+        has_slash = true;
+    }
+
+    uri = talloc_asprintf(dc_ctx, "%s%s.well-known/openid-configuration",
+                                   issuer_url, has_slash ? "" : "/");
+    if (uri == NULL) {
+        DEBUG(SSSDBG_OP_FAILURE, "Failed to allocate memory for results.\n");
+        ret = ENOMEM;
+        goto done;
+    }
+
+    clean_http_data(dc_ctx);
+    ret = do_http_request(dc_ctx, uri, NULL, NULL);
+    if (ret != EOK) {
+        DEBUG(SSSDBG_OP_FAILURE, "http request failed.\n");
+    }
+
+done:
+    talloc_free(uri);
+
+    return ret;
+}
+
+#define DEFAULT_SCOPE "user"
+
+errno_t get_devicecode(struct devicecode_ctx *dc_ctx,
+                       const char *client_id)
+{
+    int ret;
+
+    char *post_data = NULL;
+
+    post_data  = talloc_asprintf(dc_ctx, "client_id=%s&scope=%s",
+                                 client_id,
+                                 dc_ctx->scope != NULL ? dc_ctx->scope
+                                                       : DEFAULT_SCOPE);
+    if (post_data == NULL) {
+        DEBUG(SSSDBG_OP_FAILURE, "Failed to allocate memory for POST data.\n");
+        return ENOMEM;
+    }
+
+    clean_http_data(dc_ctx);
+    ret = do_http_request(dc_ctx, dc_ctx->device_authorization_endpoint,
+                          post_data, NULL);
+    if (ret != EOK) {
+        DEBUG(SSSDBG_OP_FAILURE, "Failed to send device code request.\n");
+    }
+
+    talloc_free(post_data);
+    return ret;
+}
+
+errno_t get_userinfo(struct devicecode_ctx *dc_ctx)
+{
+    int ret;
+
+    clean_http_data(dc_ctx);
+    ret = do_http_request(dc_ctx, dc_ctx->userinfo_endpoint, NULL,
+                          dc_ctx->access_token_str);
+    if (ret != EOK) {
+        DEBUG(SSSDBG_OP_FAILURE, "Failed to send userinfo request.\n");
+    }
+
+    return ret;
+}
+
+errno_t get_jwks(struct devicecode_ctx *dc_ctx)
+{
+    int ret;
+
+    clean_http_data(dc_ctx);
+    ret = do_http_request(dc_ctx, dc_ctx->jwks_uri, NULL, NULL);
+    if (ret != EOK) {
+        DEBUG(SSSDBG_OP_FAILURE, "Failed to read jwks file [%s].\n",
+                                 dc_ctx->jwks_uri);
+    }
+
+    return ret;
+
+}
+
+static int cleanup_curl(void *p)
+{
+    curl_global_cleanup();
+
+    return 0;
+}
+
+errno_t init_curl(struct devicecode_ctx *dc_ctx)
+{
+    CURLcode res;
+
+    res = curl_global_init(CURL_GLOBAL_ALL);
+    if (res != CURLE_OK) {
+        DEBUG(SSSDBG_OP_FAILURE, "Failed to initialize libcurl.\n");
+        return EIO;
+    }
+
+    if (dc_ctx != NULL) {
+        talloc_set_destructor((void *) dc_ctx, cleanup_curl);
```

Comment:
Imo, instead of a cast here it would make sense to simply supply `(void *)` directly to `init_curl()`